pkg/policy/mapstate.go

```diff
@@ -273,6 +284,224 @@ func (msm *mapStateMap) ForEach(f func(Key, MapStateEntry) bool) bool {
 	return true
 }
 
+func (msm *mapStateMap) forKey(k Key, f func(Key, MapStateEntry) bool) bool {
+	e, ok := msm.entries[k]
+	if ok {
+		return f(k, e)
+	}
+	stacktrace := hclog.Stacktrace()
+	log.Errorf("Missing MapStateEntry for key: %v. Stacktrace: %s", k, stacktrace)
+	return true
+}
+
+// ForEachNarrowerCIDRAncestor iterates over narrower port/proto's and broader CIDRs in the trie.
+// Equal port/protos or identities are not included.
+func (msm *mapStateMap) ForEachNarrowerCIDRAncestor(key Key, prefixes []netip.Prefix, f func(Key, MapStateEntry) bool) {
+	msm.trie.Descendants(key.PrefixLength(), key, func(_ uint, lpmKey bitlpm.Key[policyTypes.Key], idSet IDSet) bool {
+		// k is the key from trie with 0'ed ID
+		k := lpmKey.Value()
+
+		// Descendants iterates over the key itself if it is in the trie, skip it
+		if k.PortProtoIsEqual(key) {
```

Comment:
This check should live in the argument func. I also don't think this comment is right.

Reply:
Yes, the comment was wrong. I think the code is better here as it is needed to implement the "Narrower" semantic and we also must exclude identical identities, which makes the function structurally different.

Reply:
I'll have think again if I could name these functions better, though.

Reply:
Moving this check on the port/proto to the argument function would cause unnecessary traversals of the IDs under this port/proto in the IDSet or the CIDR trie, which are needed to call the argument function. IMO we should not do these unnecessary traversals.